### PR TITLE
feat: ability to add custom headers to S3 upload calls

### PR DESCRIPTION
**Pull request by zendesk-klross**
https://github.com/Netflix/metaflow/issues/1191

### CODE REVIEW

File: metaflow/metaflow_config_funcs.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def config_values(include=0):
             yield name, config_value.serializer(config_value.value)
 
 
-def from_conf(name, default=None, validate_fn=None):
+def from_conf(name, default=None, validate_fn=None, prefix=False):
```
[the remark this answers is not in the document]
Comment (saikonen) [reply metaflow/metaflow_config_funcs.py:52]: I'm also not entirely sure if we want to task the `from_conf` with this kind of dict-building, or if a separate function would make more sense for this use case. 
The prefix-parsing is a bit of a niche for the time being, and only works when the `key` can be used as an environment variable, so reusability for f.ex. building a dict with hyphenated keys will not work.

File: metaflow/plugins/aws/batch/batch.py [hunk 2]
```diff
@@ -262,6 +263,11 @@ def create_job(
         if tmpfs_enabled and tmpfs_tempdir:
             job.environment_variable("METAFLOW_TEMPDIR", tmpfs_path)
 
+        if S3_UPLOAD_ARGS is not None:
+            for key, value in S3_UPLOAD_ARGS.items():
+                env_var = "METAFLOW_S3_UPLOAD_ARGS_" + key
+                job.environment_variable(env_var, value)
+
```
[the remark this answers is not in the document]
Comment (zendesk-klross) [reply metaflow/plugins/aws/batch/batch.py:270]: Thanks for your feedback! We've tested this with a simple HelloAWS flow and it worked fine, but we haven't tried it with a foreach loop yet.

File: metaflow/metaflow_config_funcs.py [hunk 3]
```diff
@@ -106,6 +106,28 @@ def from_conf(name, default=None, validate_fn=None):
     return value
 
 
+def from_conf_dict(prefix, validate_fn=None):
```
Comment (savingoyal) [inline metaflow/metaflow_config_funcs.py:109]: @zendesk-klross Any specific reason for this change? Given that this PR focuses on just the encryption bits - I wonder if we need to support arbitrary headers?

Comment (savingoyal) [reply metaflow/metaflow_config_funcs.py:109]: Happy to jump on a quick call if that would be useful to get to a quick conclusion.

Comment (zendesk-klross) [reply metaflow/metaflow_config_funcs.py:109]: We added the encryption attribute to S3 objects mainly to test a use-case for the new custom headers (this way, we can make sure the object actually gets updated using these headers). But we also figured it's a good idea to let users pick their own headers for their specific needs, as mentioned [here](https://github.com/Netflix/metaflow/issues/1191).

Comment (savingoyal) [reply metaflow/metaflow_config_funcs.py:109]: We need to ensure that we take care of all scenarios with arbitrary headers, including any considerations that would affect the data being read back. If you have a specific use case, we are happy to tackle that as part of this PR.

Comment (zendesk-klross) [reply metaflow/metaflow_config_funcs.py:109]: We also thought that this parsing function might be useful for other configurable arguments that we don't want to explicitly define in Metaflow, but let us know if you can't think of any other applications for this.

Comment (zendesk-klross) [reply metaflow/metaflow_config_funcs.py:109]: Right now the PR starts the groundwork to support all these options: `ALLOWED_UPLOAD_ARGS = ['ACL', 'CacheControl', 'ChecksumAlgorithm', 'ContentDisposition', 'ContentEncoding', 'ContentLanguage', 'ContentType', 'ExpectedBucketOwner', 'Expires', 'GrantFullControl', 'GrantRead', 'GrantReadACP', 'GrantWriteACP', 'Metadata', 'ObjectLockLegalHoldStatus', 'ObjectLockMode', 'ObjectLockRetainUntilDate', 'RequestPayer', 'ServerSideEncryption', 'StorageClass', 'SSECustomerAlgorithm', 'SSECustomerKey', 'SSECustomerKeyMD5', 'SSEKMSKeyId', 'SSEKMSEncryptionContext', 'Tagging', 'WebsiteRedirectLocation'].` The list is pretty long and testing for all combinations of headers might be complicated, so if you feel that an untested setup isn't a good idea we can definitely refocus the PR to support ServerSideEncryption only.

Comment (zendesk-klross) [reply metaflow/metaflow_config_funcs.py:109]: We're also happy to expand this PR to include support for these `ALLOWED_DOWNLOAD_ARGS = ['ChecksumMode', 'VersionId', 'SSECustomerAlgorithm', 'SSECustomerKey', 'SSECustomerKeyMD5', 'RequestPayer', 'ExpectedBucketOwner']`if you think that could be useful.

Comment (savingoyal) [reply metaflow/metaflow_config_funcs.py:109]: @zendesk-klross We can make this PR specific to encryption to keep the testing burden low for now if there aren't any other motivating use cases.

Comment (savingoyal) [reply metaflow/metaflow_config_funcs.py:109]: Would you be able to create a version that just handles that so that we can kick off the review process?

Comment (zendesk-klross) [reply metaflow/metaflow_config_funcs.py]: @savingoyal I've just opened a new PR [https://github.com/Netflix/metaflow/pull/1436](url) that limits the scope of this to server-side-encryption, let me know what you think!

**Pull request closed without merging**